source/extensions/filters/http/jwt_authn/jwks_async_fetcher.h

```diff
@@ -0,0 +1,84 @@
+#pragma once
+
+#include <chrono>
+
+#include "envoy/extensions/filters/http/jwt_authn/v3/config.pb.h"
+#include "envoy/server/factory_context.h"
+
+#include "common/common/logger.h"
+#include "common/init/target_impl.h"
+
+#include "extensions/filters/http/common/jwks_fetcher.h"
+#include "extensions/filters/http/jwt_authn/stats.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace JwtAuthn {
+
+/**
+ *  CreateJwksFetcherCb is a callback interface for creating a JwksFetcher instance.
+ */
+using CreateJwksFetcherCb = std::function<Common::JwksFetcherPtr(Upstream::ClusterManager&)>;
+/**
+ *  JwksDoneFetched is a callback interface to set a Jwks when fetch is done.
+ */
+using JwksDoneFetched = std::function<void(google::jwt_verify::JwksPtr&& jwks)>;
+
+// This class handles fetching Jwks asynchronously.
+// It will be no-op if async_fetch is not enabled.
+// At its constructor, it will start to fetch Jwks, register with init_manager if not fast_listener.
+// and handle fetching response. When cache is expired, it will fetch again.
+// When a Jwks is fetched, done_fn is called to set the Jwks.
+class JwksAsyncFetcher : public Logger::Loggable<Logger::Id::jwt>,
+                         public Common::JwksFetcher::JwksReceiver {
+public:
+  JwksAsyncFetcher(const envoy::extensions::filters::http::jwt_authn::v3::RemoteJwks& remote_jwks,
+                   Server::Configuration::FactoryContext& context, CreateJwksFetcherCb fetcher_fn,
+                   JwtAuthnFilterStats& stats, JwksDoneFetched done_fn);
+
+  ~JwksAsyncFetcher();
+
+  // Get the remote Jwks cache duration.
+  static std::chrono::seconds
+  getCacheDuration(const envoy::extensions::filters::http::jwt_authn::v3::RemoteJwks& remote_jwks);
+
+private:
+  // Start to fetch Jwks
+  void refresh();
+  // Handle fetch done.
+  void handle_fetch_done();
+
+  // Override the functions from Common::JwksFetcher::JwksReceiver
+  void onJwksSuccess(google::jwt_verify::JwksPtr&& jwks) override;
+  void onJwksError(Failure reason) override;
```

Comment:
I'm not sure it's a good idea to change JwksReceiver's public interface in its implementation. Probably worth a comment on why it's done this way.

Reply:
Not sure what change you like to have?

Reply:
These two methods are made private in the derived class. Is there a reason for that?

Reply:
These two are virtual function override.  They are not supposedly to be called directly by others, hence marking them private. 
Their base class interfaces are public.

Reply:
Quoted from from [here](http://www.gotw.ca/publications/mill18.htm)
`Guideline #2: Prefer to make virtual functions private.`

Reply:
The quote is about the "Template Method" pattern which we don't use in Envoy. For example the [C++ FAQ](https://isocpp.org/wiki/faq/proper-inheritance#hiding-inherited-public) has a different opinion on that.